Development and testing of Ansible roles and playbooks

Durf1985 opened the pull request:
Creating local test infrastracture for playbooks and roles

## In the process done:

* Local development using Vagrant, refinement of roles for provisioning in Vagrant
 * Role testing using Molecule and Testinfra
 * Switching image collection by packer to using roles
 

## How to setup the project:
* In directory ~/your_repo/ansible/ execute following command:
```bash
`vagrant up`
```
after that create molecule testing infrastracture
```bash
molecule init scenario -r your_role_name -d vagrant your_scenario_name
```
add additional tests if it necessary, and configure the converge.yml and molecule.yml in created directory 

## How to check the project:


* In directory ~/your_repo/ansible/roles/your_role_directory execute following command:
```bash
molecule test
```

## PR checklist

The label with the homework number, in label description - homework lecture and topic

Durf1985 committed:
DO-90 Create Vagrantfile to provision app/db serv

Created Vagrantfile config

Durf1985 committed:
DO-91 Add provisioners to Vagrantfile

Added provisioners to Vagrantfile.
Reconfigured tasks in roles (app,db), divided big tasks to many sub-tasks
Created and reconfigured templates
Reduced hardcode in tasks

loktionovam commented on ansible/playbooks/users.yml:
```diff
@@ -3,10 +3,13 @@
+    # - "{{ inventory_dir }}/credentials.yml"
 
   tasks:
+    - name: Debug
```
This is unnecessary and should be deleted

Durf1985 committed:
DO-92 Configure nginx in Vagrant

Configured Vagrantfile for reddit-app via nginx

Durf1985 committed:
DO-93 Molecule to test db role

Add and  configured molecule in db role

loktionovam commented on ansible/roles/db/molecule/default/INSTALL.rst:
```diff
@@ -0,0 +1,23 @@
+
+.. code-block:: bash
+
+    $ pip install 'molecule_vagrant'
```
It is better to move this to requirements.txt file and pin the version.

Durf1985 replied:
I move molecule_vagrant to requiremnts.txt, but I did not touch this file, i.e. delete it, or just clear it, or some other option

loktionovam commented on ansible/roles/db/molecule/default/molecule.yml:
```diff
@@ -0,0 +1,17 @@
+  name: vagrant
+  provider:
+    name: virtualbox
+lint: yamllint
```
I've got the error:

```bash
╰─❯ molecule test                                                                                                                                                                                       ─╯
INFO     default scenario test matrix: dependency, lint, cleanup, destroy, syntax, create, prepare, converge, idempotence, side_effect, verify, cleanup, destroy
INFO     Performing prerun with role_name_check=0...
INFO     Set ANSIBLE_LIBRARY=/home/isaak/.cache/ansible-compat/7bdc25/modules:/home/isaak/.ansible/plugins/modules:/usr/share/ansible/plugins/modules
INFO     Set ANSIBLE_COLLECTIONS_PATH=/home/isaak/.cache/ansible-compat/7bdc25/collections:/home/isaak/.ansible/collections:/usr/share/ansible/collections
INFO     Set ANSIBLE_ROLES_PATH=/home/isaak/.cache/ansible-compat/7bdc25/roles:/home/isaak/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles
INFO     Running default > dependency
WARNING  Skipping, missing the requirements file.
WARNING  Skipping, missing the requirements file.
INFO     Running default > lint
CRITICAL Deprecated linter config found, migrate to v3 schema. See https://github.com/ansible-community/molecule/issues/2293
```

Migrate your linter configuration or repin your packages in requirements.txt with appropriate version. The first is preferably

loktionovam commented on ansible/Vagrantfile:
```diff
@@ -0,0 +1,35 @@
+          "app" => ["tag_reddit-app"],
+          "app:vars" => {"db_host" => "192.168.56.10"},
+        }
+        ansible.extra_vars = "/home/fallgeratoor/github_repo/devops_study_infra/ansible/environments/dev/group_vars/tag_reddit-app.yml"
```
Do not use absolute paths because this code is not reusable and portable

loktionovam commented on ansible/ansible.cfg:
```diff
@@ -14,3 +14,4 @@ vault_password_file = ~/.ansible/vault.key
 [diff]
 always = True
 context = 5
+deb [arch=amd64 signed-by=/usr/share/keyrings/oracle-virtualbox-2016.gpg] https://download.virtualbox.org/virtualbox/debian <mydist> contrib
```
What does this string mean?

loktionovam commented on ansible/playbooks/site.yml:
```diff
@@ -1,5 +1,8 @@
+  ansible.builtin.import_playbook: app.yml
+- name: Install Reddit application
+  ansible.builtin.import_playbook: deploy.yml
+# - ansible.builtin.import_playbook: users.yml
```
If you want to remove your code it is better to remove it rather than comment because you can accidentally uncomment this and get an error

loktionovam commented on ansible/roles/db/molecule/default/tests/test_default.py:
```diff
@@ -0,0 +1,23 @@
+testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(
+    os.environ['MOLECULE_INVENTORY_FILE']).get_hosts('all')
+
+# check if MongoDB is enabled and running
```
It is better to not write "obvious" comments (as the rule: add comments only where they are strictly necessary)

loktionovam reviewed (changes requested):
Update README.md accordingly your changes

Durf1985 committed:
DO-94 Use github actions to test your db role

Durf1985 committed:
Updating documentation

Added documentation for lectures from 10 to 13 in README.md
Minor changed PULL_REQUEST_TEMPLATE

loktionovam commented on README.md:
```diff
@@ -775,3 +774,202 @@ resource "google_compute_global_forwarding_rule" "app" {
+echo "deb [signed-by=/usr/share/keyrings/hashicorp-archive-keyring.gpg] https://apt.releases.hashicorp.com $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/hashicorp.list
+sudo apt update && sudo apt install vagrant
+
+echo "deb [arch=amd64 signed-by=/usr/share/keyrings/oracle-virtualbox-2016.gpg] https://download.virtualbox.org/virtualbox/debian bionic contrib" | sudo tee -a /etc/apt/sources.list.d/virtualbox.list
```
It is better to use `$(lsb_release -cs)` rather than hardcode `bionic`

loktionovam commented on README.md:
```diff
@@ -775,3 +774,202 @@ resource "google_compute_global_forwarding_rule" "app" {
+
+* Create GCE Service Account <https://cloud.google.com/compute/docs/access/service-accounts> and save on your local host in directory `~/your_repo/ansible/gce_py/` key in `*.json` format from GCP for your dynamic inventory script
+  
+* In directory `environments` contains files and folders responsible for various environments, which contain their own variables and settings, as well as a dynamic inventory script. Your roles will take values from these directories depending on the environment
```
`contains files and folders responsible` -> `place files and folders responsible`

loktionovam commented on README.md:
````diff
@@ -775,3 +774,202 @@ resource "google_compute_global_forwarding_rule" "app" {
+   
+```
+
+* To check your dinamic inventory execute in directory `~/your_repo/ansible` next command
````
`dinamic` -> `dynamic`

loktionovam commented on README.md:
````diff
@@ -775,3 +774,202 @@ resource "google_compute_global_forwarding_rule" "app" {
+```
+
+* You place your playbooks in the `playbooks` directory. In the `roles` directory you place the roles
+* Roles have next sturcture
````
`next` -> `the following`

loktionovam commented on README.md:
````diff
@@ -775,3 +774,202 @@ resource "google_compute_global_forwarding_rule" "app" {
+ansible-galaxy install -r environments/stage/requirements.yml  
+```
+
+* To launch playbooks, run the following command in the direcotry `~/your_repo/ansible`
````
`launch playbooks` -> `run playbooks`, `direcotry` -> `directory`
